metricbeat/module/kubernetes/util/enrichers.md

```diff
@@ -0,0 +1,112 @@
+- [kubernetes.Init](https://github.com/constanca-m/beats/blob/8fcdfc205ebe0221ed7a1046c673c9babfb5280d/metricbeat/helper/kubernetes/state_metricset.go#L44) calls the [New metricsets method](https://github.com/constanca-m/beats/blob/8fcdfc205ebe0221ed7a1046c673c9babfb5280d/metricbeat/helper/kubernetes/state_metricset.go#L80), which returns a new metricset with resource metadata enricher as part of it.
+- Node, pod, container, and `state_container` metricsets do not trigger the `kubernetes.Init` rather they implement their own [New method](https://github.com/constanca-m/beats/blob/8fcdfc205ebe0221ed7a1046c673c9babfb5280d/metricbeat/module/kubernetes/node/node.go#L66) with the enricher as part of the metricsets returned.
+- All of the above metricsets trigger the [NewResourceMetadataEnricher](https://github.com/constanca-m/beats/blob/8fcdfc205ebe0221ed7a1046c673c9babfb5280d/metricbeat/helper/kubernetes/state_metricset.go#L84) with only the exception of the container and `state_container` metricsets which trigger the [NewContainerMetadataEnricher](https://github.com/constanca-m/beats/blob/8fcdfc205ebe0221ed7a1046c673c9babfb5280d/metricbeat/module/kubernetes/state_container/state_container.go#L118C23-L118C51).
+- The [NewResourceMetadataEnricher](https://github.com/constanca-m/beats/blob/8fcdfc205ebe0221ed7a1046c673c9babfb5280d/metricbeat/module/kubernetes/util/kubernetes.go#L501) is the function responsible for creating and returning a metadata enricher.
+- The enricher is responsible for the metadata enrichment. For that, resource watchers are used which are shared between the different metricsets. For example for pod metricset, a pod watcher, a namespace and node watcher are by default needed in addition to job and replicaset watcher based on the configuration. These watchers will be also used by other metricsets that require them
+like state_pod, state_container, node, state_job etc.
+- The shared watchers are stored in the module-level struct [resourceWatchers](https://github.com/constanca-m/beats/blob/8fcdfc205ebe0221ed7a1046c673c9babfb5280d/metricbeat/module/kubernetes/kubernetes.go#L102). The struct is initialized when the Kubernetes module [starts](https://github.com/constanca-m/beats/blob/8fcdfc205ebe0221ed7a1046c673c9babfb5280d/metricbeat/module/kubernetes/util/kubernetes.go#L124).
+- Each time a watcher gets created it is added to the struct along with some watcher important data. The [struct](https://github.com/constanca-m/beats/blob/8fcdfc205ebe0221ed7a1046c673c9babfb5280d/metricbeat/module/kubernetes/util/kubernetes.go#L99) consists of a metaWatchersMap which has as key the name of the resource this watcher is watching(i.e. pod or node) and as value a [metaWatcher](https://github.com/constanca-m/beats/blob/8fcdfc205ebe0221ed7a1046c673c9babfb5280d/metricbeat/module/kubernetes/util/kubernetes.go#L86C6-L86C17) struct. 
+- The metaWatcher struct holds all the relevant info for the specific watcher. These info include
```

Comment:
These details of the variables are already in the code like this:

```
	watcher kubernetes.Watcher
	started bool // true if watcher has started, false otherwise

	metricsetsUsing []string // list of metricsets using this watcher

	enrichers       map[string]*enricher // map of enrichers using this watcher. The key is the metricset name
	metadataObjects map[string]bool      // map of ids of each object received by the handler functions

	nodeScope      bool               // whether this watcher is only for current node
	restartWatcher kubernetes.Watcher // whether this watcher needs a restart
```